### DISCUSSION TITLE
DataGrid SelectedItem unexpected change after SourceCache AddOrUpdate call

### DISCUSSION THREAD

**Discussion by CzBuCHi** [AvaloniaUI/Avalonia#15214] [Q&A]
### Describe the bug

i have `SourceCache<DataModel, string> ProjectsCache` that is connected to ui:

xaml
```
<DataGrid ItemsSource="{Binding Projects}" SelectedItem="{Binding SelectedProject}">
    <DataGrid.Columns>
        <DataGridTextColumn Header="Identifier" Binding="{Binding Identifier}" />
        <DataGridTextColumn Header="Value" Binding="{Binding Value}" />
    </DataGrid.Columns>
</DataGrid>
```

viewmodel
```
private ReadOnlyObservableCollection<DataViewModel>? _Projects;
public ReadOnlyObservableCollection<DataViewModel> Projects {
    get {
        if (_Projects == null) {
            ProjectsCache.Connect()
                .Transform(o => new DataViewModel { Identifier = o.Identifier, Value = o.Value })
                .Sort(SortExpressionComparer<DataViewModel>.Ascending(o => o.Identifier))
                .Bind(out _Projects)
                .DisposeMany()
                .Subscribe();
        }

        return _Projects!;
    }
}
``` 
when i call
```
ProjectsCache.Edit(o =>{
    var first = o.Lookup("First");
    first.Value.Value = "VALUE";
    o.AddOrUpdate(first.Value);
});
```
DataGrid SelectedItem get changed from [ExternalCode] - from stack trace it shown as result of AddOrUpdate call

### To Reproduce

Minimal code to reproduce: 
[AvaloniaApplication.zip](https://github.com/AvaloniaUI/Avalonia/files/14850638/AvaloniaApplication.zip)

### Expected behavior

Grid selected row (aka SelectedIndex / SelectedItem) do not change but column value do update.

### Avalonia version

11.0.10

### OS

Windows

**Comment by timunie** [comment]
missing line

```diff
    ProjectsCache.Connect()
        .Transform(o => new DataViewModel { Identifier = o.Identifier, Value = o.Value })
        .Sort(SortExpressionComparer<DataViewModel>.Ascending(o => o.Identifier))
+       .ObserveOn(RxApp.MainThreadScheduler) // Needed to make sure events raised on UIThread
        .Bind(out _Projects)
        .DisposeMany()
        .Subscribe();
```

**Reply by CzBuCHi** [reply]
after adding your line SelectedItem is still changed ...

**Reply by CzBuCHi** [reply]
@timunie just noticed, that you changed label to by-design which i do not understand ...

why update of background model trigger change of selected row on view? (imagine situation where background worker is periodically updating ProjectsCache items - then selected row on DataGrid would 'randomly' change without user interaction ...)

**Reply by timunie** [reply]
well... The UI requires the update events to be raised from UI Thread. That doesn't mean you can't fetch the ihtems on a worker thread. But the event must happen on the UIThread. The line I added makes sure it does redirect the events as needed...

**Reply by CzBuCHi** [reply]
but that not my problem (btw: ui did refresh even without ObserveOn call) ... 

in my example project i have 2 items in ProjectsCache:   `{ Identifier: "First", Value: "" }` and `{ Identifier: "Second", Value:  "" }`
so DataGrid shows me table with 2 rows: ('>' represents selected row)
||Identifier|Value|
|-|-|-|
|>|First||
||Second||

now when this code is executed:
```
var identifier = "First"; // passed as argument
ProjectsCache.Edit(o => {
    var first = o.Lookup(identifier);
    first.Value.Value = "VALUE";
    o.AddOrUpdate(first.Value);
});
```
that causes table to refresh first row (expected), but also change selected row to second (unexpected):
||Identifier|Value|
|-|-|-|
||First|VALUE|
|>|Second||

but i need to keep selected row to first one:

||Identifier|Value|
|-|-|-|
|>|First|VALUE|
||Second||

**Reply by timunie** [reply]
`first.Value.Value = "VALUE";` should be enough to update Value here if your RowItem implements `INotifyPropertyChanged` which I highly suggest. AddOrUpdate will remove the row and add it back. For sure DataGrid cannot select a row that was removed.
